File: backend/app/core/nickname.py
"""Nickname validation — single source of truth for backend rules."""

from __future__ import annotations

NICKNAME_LENGTH_MSG = "От 2 до 32 символов"
NICKNAME_DIGITS_MSG = "Ник не может состоять только из цифр"
NICKNAME_TAKEN_MSG = "Ник занят"

NICKNAME_MIN_LEN = 2
NICKNAME_MAX_LEN = 32


def normalize_nickname_key(value: str) -> str:
    """Casefold key for uniqueness comparison (store original separately)."""
    return value.casefold()


def validate_nickname(value: str) -> str:
    """Strip and validate; return original casing. Raises ValueError with RU message.

    Any Unicode is allowed (letters, spaces, punctuation, emoji). Constraints:
    length 2–32 after strip, and not digits-only.
    """
    cleaned = value.strip()
    if len(cleaned) < NICKNAME_MIN_LEN or len(cleaned) > NICKNAME_MAX_LEN:
        raise ValueError(NICKNAME_LENGTH_MSG)
    if cleaned.isdigit():
        raise ValueError(NICKNAME_DIGITS_MSG)
    return cleaned

File: frontend/src/lib/nickname.ts
import { z } from "zod";

/** Any Unicode (letters, spaces, punctuation, emoji). Length + not digits-only. */
export const NICKNAME_LENGTH_MSG = "От 2 до 32 символов";
export const NICKNAME_DIGITS_MSG = "Ник не может состоять только из цифр";
export const NICKNAME_TAKEN_MSG = "Ник занят";
export const NICKNAME_HINT = "Любые символы, в том числе пробелы и emoji";

export const NICKNAME_MIN_LEN = 2;
export const NICKNAME_MAX_LEN = 32;

export function isNicknameDigitsOnly(value: string): boolean {
  return /^\d+$/.test(value);
}

/** Zod field — same rules as backend `app.core.nickname`. */
export const nicknameFieldSchema = z
  .string()
  .trim()
  .min(NICKNAME_MIN_LEN, NICKNAME_LENGTH_MSG)
  .max(NICKNAME_MAX_LEN, NICKNAME_LENGTH_MSG)
  .refine((value) => !isNicknameDigitsOnly(value), NICKNAME_DIGITS_MSG);

export const nicknameFormSchema = z.object({
  nickname: nicknameFieldSchema,
});
